sync2/p2p.go

```diff
@@ -0,0 +1,211 @@
+	MinCompleteFraction    float64       `mapstructure:"min-complete-fraction"`
+	SplitSyncGracePeriod   time.Duration `mapstructure:"split-sync-grace-period"`
+	RecentTimeSpan         time.Duration `mapstructure:"recent-time-span"`
+	EnableActiveSync       bool          `mapstructure:"enable-active-sync"`
+	MaxReconcDiff          float64       `mapstructure:"max-reconc-diff"`
+	AutoCommitCount        int           `mapstructure:"auto-commit-count"`
+	AutoCommitIdle         time.Duration `mapstructure:"auto-commit-idle"`
+	TrafficLimit           int           `mapstructure:"traffic-limit"`
+	MessageLimit           int           `mapstructure:"message-limit"`
+}
+
+// DefaultConfig returns the default configuration for the P2PHashSync.
+func DefaultConfig() Config {
+	return Config{
+		MaxSendRange:           rangesync.DefaultMaxSendRange,
+		SampleSize:             rangesync.DefaultSampleSize,
+		SyncPeerCount:          20,
+		MinSplitSyncPeers:      2,
+		MinSplitSyncCount:      1000,
+		MaxFullDiff:            10000,
+		SyncInterval:           5 * time.Minute,
+		MinCompleteFraction:    0.5,
+		SplitSyncGracePeriod:   time.Minute,
+		NoPeersRecheckInterval: 30 * time.Second,
+		MaxReconcDiff:          0.01,
+		AutoCommitCount:        10000,
+		AutoCommitIdle:         time.Second,
+		TrafficLimit:           200_000_000,
+		MessageLimit:           20_000_000,
+	}
+}
+
+// P2PHashSync is handles the synchronization of a local OrderedSet against other peers.
+type P2PHashSync struct {
+	logger     *zap.Logger
+	cfg        Config
+	os         multipeer.OrderedSet
+	syncBase   multipeer.SyncBase
+	reconciler *multipeer.MultiPeerReconciler
+	cancel     context.CancelFunc
+	eg         errgroup.Group
+	start      sync.Once
+	running    atomic.Bool
+}
+
+// NewP2PHashSync creates a new P2PHashSync.
+func NewP2PHashSync(
+	logger *zap.Logger,
+	d *Dispatcher,
+	name string,
+	os multipeer.OrderedSet,
+	keyLen, maxDepth int,
+	peers *peers.Peers,
+	handler multipeer.SyncKeyHandler,
+	cfg Config,
+	requester rangesync.Requester,
+) *P2PHashSync {
+	s := &P2PHashSync{
+		logger: logger,
+		os:     os,
+		cfg:    cfg,
+	}
+	rangeSyncOpts := []rangesync.RangeSetReconcilerOption{
+		rangesync.WithMaxSendRange(cfg.MaxSendRange),
+		rangesync.WithSampleSize(cfg.SampleSize),
+		rangesync.WithMaxDiff(cfg.MaxReconcDiff),
+		rangesync.WithLogger(logger),
+	}
+	if cfg.RecentTimeSpan > 0 {
+		rangeSyncOpts = append(rangeSyncOpts, rangesync.WithRecentTimeSpan(cfg.RecentTimeSpan))
+	}
+	// var ps multipeer.PairwiseSyncer
+	ps := rangesync.NewPairwiseSetSyncer(requester, name, rangeSyncOpts, []rangesync.ConduitOption{
+		rangesync.WithTrafficLimit(cfg.TrafficLimit),
+		rangesync.WithMessageLimit(cfg.MessageLimit),
+	})
+	s.syncBase = multipeer.NewSetSyncBase(ps, s.os, handler)
+	s.reconciler = multipeer.NewMultiPeerReconciler(
+		s.syncBase, peers, keyLen, maxDepth,
+		multipeer.WithLogger(logger),
+		multipeer.WithSyncPeerCount(cfg.SyncPeerCount),
```

Comment:
seems like a good candidate for a `Config` type. many function calls that can be avoided

Reply:
Switched to config type (also did that for `rangesync` package)